examples/conversation_with_agent_with_finetuned_model/huggingface_model.py

```diff
@@ -0,0 +1,509 @@
+        except Exception as e:
+            logger.error(f"Generation error: {e}")
+            raise
+
+    def format(
+        self,
+        *args: Union[MessageBase, Sequence[MessageBase]],
+    ) -> List[dict]:
+        """A basic strategy to format the input into the required format of
+        Hugging Face models.
+
+        Args:
+            args (`Union[MessageBase, Sequence[MessageBase]]`):
+                The input arguments to be formatted, where each argument
+                should be a `Msg` object, or a list of `Msg` objects.
+                In distribution, placeholder is also allowed.
+
+        Returns:
+            `List[dict]`:
+                The formatted messages.
+        """
+        huggingface_msgs = []
+        for msg in args:
+            if msg is None:
+                continue
+            if isinstance(msg, MessageBase):
+                # content shouldn't be empty string
+                if msg.content == "":
+                    logger.warning(
+                        "The content field cannot be "
+                        "empty string. To avoid error, the empty string is "
+                        "replaced by a blank space automatically, but the "
+                        "model may not work as expected.",
+                    )
+                    msg.content = " "
+
+                huggingface_msg = {
+                    "role": msg.role,
+                    "content": _convert_to_str(msg.content),
+                }
+
+                # image url
+                if msg.url is not None:
+                    huggingface_msg["images"] = [msg.url]
+
+                huggingface_msgs.append(huggingface_msg)
+            elif isinstance(msg, list):
+                huggingface_msgs.extend(self.format(*msg))
+            else:
+                raise TypeError(
+                    f"Invalid message type: {type(msg)}, `Msg` is expected.",
+                )
+
+        return huggingface_msgs
+
+    def load_model(
+        self,
+        model_id: Optional[str] = None,
+        local_model_path: Optional[str] = None,
+    ) -> None:
+        """
+        Load a new model for the agent from
+        a local path and update the agent's model.
+
+        Arguments:
+            local_model_path (str): The file path to the model to be loaded.
+            model_id (str): An identifier for the model on Huggingface.
+
+        Raises:
+            Exception: If the model cannot be loaded from the given
+                       path or identifier.
+                       Possible reasons include file not found,
+                       incorrect model ID,
+                       or network issues while fetching the model.
+        """
+
+        try:
+            if local_model_path is None:
+                self.model = AutoModelForCausalLM.from_pretrained(
+                    model_id,
+                    token=self.huggingface_token,
+                    device_map="auto",
```

Comment:
I presumed fine-tuning of LLMs happens on gpus, since it is (as far as I'm aware) infeasible to do on gpu. `self.device` is meant for device for inference only.

Reply:
This part is updated. If the user didn't specify `device` in `model_configs`, `device_map="auto"` by default; otherwise `device_map` is set to the user-specified `device`. (see [https://github.com/modelscope/agentscope/pull/240/commits/1d0704d544e80d30ecd5d12db35a64c73f0a60af](https://github.com/modelscope/agentscope/pull/240/commits/1d0704d544e80d30ecd5d12db35a64c73f0a60af))